pkg/chunk/chunk_store.go

```diff
@@ -482,3 +482,109 @@ func (c *store) convertChunkIDsToChunks(ctx context.Context, userID string, chun
 
 	return chunkSet, nil
 }
+
+func (c *store) DeleteChunk(ctx context.Context, from, through model.Time, userID, chunkID string, metric labels.Labels, partiallyDeletedInterval *model.Interval) error {
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return errors.New("No metric name label")
+	}
+
+	chunkWriteEntries, err := c.schema.GetWriteEntries(from, through, userID, string(metricName), metric, chunkID)
+	if err != nil {
+		return err
+	}
+
+	return c.deleteChunk(ctx, userID, chunkID, metric, chunkWriteEntries, partiallyDeletedInterval, func(chunk Chunk) error {
+		return c.PutOne(ctx, chunk.From, chunk.Through, chunk)
+	})
+}
+
+func (c *store) deleteChunk(ctx context.Context, userID, chunkID string, metric labels.Labels,
+	chunkWriteEntries []IndexEntry, partiallyDeletedInterval *model.Interval, putChunkFunc func(chunk Chunk) error) error {
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return errors.New("No metric name label")
+	}
+
+	// if chunk is partially deleted, fetch it, slice non-deleted portion and put it to store before deleting original chunk
+	if partiallyDeletedInterval != nil {
+		err := c.deleteChunkPartially(ctx, userID, chunkID, *partiallyDeletedInterval, putChunkFunc)
+		if err != nil {
+			return err
+		}
+	}
+
+	batch := c.index.NewWriteBatch()
+	for i := range chunkWriteEntries {
+		batch.Delete(chunkWriteEntries[i].TableName, chunkWriteEntries[i].HashValue, chunkWriteEntries[i].RangeValue)
+	}
+
+	err := c.index.BatchWrite(ctx, batch)
+	if err != nil {
+		return err
+	}
+
+	return c.chunks.DeleteChunk(ctx, chunkID)
+}
+
+func (c *store) deleteChunkPartially(ctx context.Context, userID, chunkID string, partiallyDeletedInterval model.Interval, putChunkFunc func(chunk Chunk) error) error {
```

Comment:
I think this function should be renamed. It doesn't partially delete a chunk, instead it creates a chunk outside of the deletion boundary. Consider calling it `sliceChunk` or `reboundChunk`.

Reply:
`sliceChunk` sounds good to me. Thanks!

Reply:
I think I will go with `reboundChunk` because it creates a confusion with `Chunk.Slice` method which creates a new chunk for the given interval while this removes the given interval from the chunk.